ocaml/lambda/matching.ml

```diff
@@ -1627,14 +1660,27 @@ and precompile_or ~arg ~arg_sort (cls : Simple.clause list) ors args def k =
                 default = Default_environment.pop_compat orp def
               }
             in
-            let pm_fv = pm_free_variables orpm in
-            let patbound_action_vars =
-              (* variables bound in the or-pattern
-                 that are used in the orpm actions *)
+            let patbound_idents =
               Typedtree.pat_bound_idents_full arg_sort orp
-              |> List.filter (fun (id, _, _, _) -> Ident.Set.mem id pm_fv)
-              |> List.map (fun (id, _, ty, id_sort) ->
-                     (id, Typeopt.layout orp.pat_env orp.pat_loc id_sort ty))
+            in
+            (* Optimization: discard pattern vars not bound in orpm actions *)
+            let patbound_idents =
+              match pm_free_variables orpm with
+              (* Give up on the optimization: there is some action not tracking
+                 free variables, so the free variable set is not known. *)
+              | Pmfv_unknown -> patbound_idents
+              (* The free variables set is known: apply the optimization by
+                 filtering out pattern-bound variables unused by the actions. *)
```

Comment:
Can you leave a CR-soon to investigate/think about this more? I don't think we're confident enough about the nature of this or-pattern optimization to say for sure whether this approach is ok (though it helps that we think we're only affecting new code, not old code). E.g. at some point I think we should identify a program where this optimization fires and reason through whether it's important.

Reply:
Added one!